### PR TITLE
f+1 validators always approve

### PR DESCRIPTION
**Pull request by rphmeier**
Any set of f+1 validators will always count as an approval on a candidate, as it contains at least one honest approval.

In small testnets, like Rococo, this can prevent failures to approve candidates when all non-backers are in tranche 0 and at least 1 no-show occurs.

### CODE REVIEW

File: node/core/approval-voting/src/approval_checking.rs [hunk 1]
```diff
@@ -119,6 +131,7 @@ pub fn check_approval(
 			if n_approved + tolerated_missing >= n_assigned {
 				Check::Approved(tolerated_missing)
 			} else {
+				println!("{} + {} < {}", n_approved, tolerated_missing, n_assigned);
```
Comment (pepyakin) [inline node/core/approval-voting/src/approval_checking.rs:134]: Stray println?

File: node/core/approval-voting/src/lib.rs [hunk 2]
```diff
@@ -1316,7 +1316,8 @@ fn check_and_apply_full_approval(
 			required_tranches.clone(),
 		);
 
-		if let approval_checking::Check::Approved(no_shows) = check {
+		let no_shows = check.known_no_shows();
+		if check.is_approved() {
```
Comment (pepyakin) [inline node/core/approval-voting/src/lib.rs:1320]: nit: why not move it under the if?

File: roadmap/implementers-guide/src/node/approval/approval-voting.md [hunk 3]
```diff
@@ -360,8 +360,9 @@ Likewise, when considering how many tranches to take, the no-show depth should b
 #### Check Approval
   * Check whether a candidate is approved under a particular block.
   * Requires `(block_entry, candidate_entry, approval_entry, n_tranches)`
+  * If we have `3 * n_approvals > 2 * n_validators`, return true. This is because any set with f+1 validators must have at least one honest validator, who has approved the candidate.
```
Comment (pepyakin) [inline roadmap/implementers-guide/src/node/approval/approval-voting.md:363]: why *2?

**Comment by burdges** [comment]
At some point, we want a cut off in which we conclude block would never be approved if we reach say 128 assignments.  In essence, this says we abort contentious candidate, rewind the relay chain, and move forward without the contentious candidate.  We'd still complete existing approval checks on the contentious candidate and slash them if found invalid.

**Review by pepyakin — APPROVED** [review]
LGTM now

File: node/core/approval-voting/src/approval_checking.rs [hunk 4]
```diff
@@ -74,6 +76,15 @@ impl Check {
...
 			Check::Approved(_) => true,
+			Check::ApprovedOneThird => true,
+		}
+	}
+
+	/// The number of known no-shows in this computation.
+	pub fn known_no_shows(&self) -> usize {
+		match *self {
+			Check::Approved(n) => n,
+			_ => 0,
```
Comment (drahnr) [inline node/core/approval-voting/src/approval_checking.rs:87]:
```suggestion
	pub fn known_no_shows(&self) -> Option<std::num::NonZeroUsize> {
		match *self {
			Check::Approved(n) => std::num::NonZeroUsize::new(n),
			_ => None,
```

there seems to be exactly one use cases which treats `0` as special case, which really should be a `None`.

Comment (rphmeier) [reply node/core/approval-voting/src/approval_checking.rs:87]: Why NonZero?

Comment (rphmeier) [reply node/core/approval-voting/src/approval_checking.rs:87]: The option makes sense, but tbh we will always use it as unwrap_or(0), so this change would add more friction

Comment (rphmeier) [reply node/core/approval-voting/src/approval_checking.rs:87]: The 'right' change would be to expose 'covered + covering' in the other RequiredTranches variants

**Pull request merged**